File: java/src/main/java/org/rocksdb/SstFileMetaData.java
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).

package org.rocksdb;

/**
 * The metadata that describes a SST file.
 */
public class SstFileMetaData extends RocksObject {
  /**
   * Called from JNI C++ or subclass
   * @param nativeHandle pointer to the C++ object
   */
  public SstFileMetaData(final long nativeHandle) {
    super(nativeHandle);
  }

  public SstFileMetaData() {
    super(newSstFileMetaData());
  }

  /**
   * Get the name of the file.
   *
   * @return the name of the file.
   */
  public String fileName() {
    return fileName(nativeHandle_);
  }

  /**
   * Get the full path where the file locates.
   *
   * @return the full path
   */
  public String path() {
    return path(nativeHandle_);
  }

  /**
   * Get the file size in bytes.
   *
   * @return file size
   */
  public long size() {
    return size(nativeHandle_);
  }

  /**
   * Get the smallest sequence number in file.
   *
   * @return the smallest sequence number
   */
  public long smallestSeqno() {
    return smallestSeqno(nativeHandle_);
  }

  /**
   * Get the largest sequence number in file.
   *
   * @return the largest sequence number
   */
  public long largestSeqno() {
    return largestSeqno(nativeHandle_);
  }

  /**
   * Get the smallest user defined key in the file.
   *
   * @return the smallest user defined key
   */
  public byte[] smallestKey() {
    return smallestKey(nativeHandle_);
  }

  /**
   * Get the largest user defined key in the file.
   *
   * @return the largest user defined key
   */
  public byte[] largestKey() {
    return largestKey(nativeHandle_);
  }

  /**
   * Get the number of times the file has been read.
   *
   * @return the number of times the file has been read
   */
  public long numReadsSampled() {
    return numReadsSampled(nativeHandle_);
  }

  /**
   * Returns true if the file is currently being compacted.
   *
   * @return true if the file is currently being compacted, false otherwise.
   */
  public boolean beingCompacted() {
    return beingCompacted(nativeHandle_);
  }

  /**
   * Get the number of entries.
   *
   * @return the number of entries.
   */
  public long numEntries() {
    return numEntries(nativeHandle_);
  }

  /**
   * Get the number of deletions.
   *
   * @return the number of deletions.
   */
  public long numDeletions() {
    return numDeletions(nativeHandle_);
  }

  private static native long newSstFileMetaData();
  @Override protected native void disposeInternal(final long handle);

  private native String fileName(final long handle);
  private native String path(final long handle);
  private native long size(final long handle);
  private native long smallestSeqno(final long handle);
  private native long largestSeqno(final long handle);
  private native byte[] smallestKey(final long handle);
  private native byte[] largestKey(final long handle);
  private native long numReadsSampled(final long handle);
  private native boolean beingCompacted(final long handle);
  private native long numEntries(final long handle);
  private native long numDeletions(final long handle);
}

File: java/rocksjni/sst_file_metadata.cc
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
//  This source code is licensed under both the GPLv2 (found in the
//  COPYING file in the root directory) and Apache 2.0 License
//  (found in the LICENSE.Apache file in the root directory).
//
// This file implements the "bridge" between Java and C++ for SstFileMetaData

#include <jni.h>

#include "include/org_rocksdb_SstFileMetaData.h"
#include "rocksdb/metadata.h"
#include "rocksjni/portal.h"

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    newSstFileMetaData
 * Signature: ()J
 */
jlong Java_org_rocksdb_SstFileMetaData_newSstFileMetaData(JNIEnv*, jclass) {
  auto* metadata = new ROCKSDB_NAMESPACE::SstFileMetaData();
  return reinterpret_cast<jlong>(metadata);
}

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    disposeInternal
 * Signature: (J)V
 */
void Java_org_rocksdb_SstFileMetaData_disposeInternal(JNIEnv*, jobject,
                                                      jlong jhandle) {
  auto* metadata =
      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileMetaData*>(jhandle);
  delete metadata;
}

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    fileName
 * Signature: (J)Ljava/lang/String;
 */
jstring Java_org_rocksdb_SstFileMetaData_fileName(JNIEnv* env, jobject,
                                                  jlong jhandle) {
  auto* metadata =
      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileMetaData*>(jhandle);
  return ROCKSDB_NAMESPACE::JniUtil::toJavaString(env, &metadata->name, false);
}

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    path
 * Signature: (J)Ljava/lang/String;
 */
jstring Java_org_rocksdb_SstFileMetaData_path(JNIEnv* env, jobject,
                                              jlong jhandle) {
  auto* metadata =
      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileMetaData*>(jhandle);
  return ROCKSDB_NAMESPACE::JniUtil::toJavaString(env, &metadata->db_path,
                                                  false);
}

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    size
 * Signature: (J)J
 */
jlong Java_org_rocksdb_SstFileMetaData_size(JNIEnv*, jobject, jlong jhandle) {
  auto* metadata =
      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileMetaData*>(jhandle);
  return static_cast<jlong>(metadata->size);
}

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    smallestSeqno
 * Signature: (J)J
 */
jlong Java_org_rocksdb_SstFileMetaData_smallestSeqno(JNIEnv*, jobject,
                                                     jlong jhandle) {
  auto* metadata =
      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileMetaData*>(jhandle);
  return static_cast<jlong>(metadata->smallest_seqno);
}

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    largestSeqno
 * Signature: (J)J
 */
jlong Java_org_rocksdb_SstFileMetaData_largestSeqno(JNIEnv*, jobject,
                                                    jlong jhandle) {
  auto* metadata =
      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileMetaData*>(jhandle);
  return static_cast<jlong>(metadata->largest_seqno);
}

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    smallestKey
 * Signature: (J)[B
 */
jbyteArray Java_org_rocksdb_SstFileMetaData_smallestKey(JNIEnv* env, jobject,
                                                        jlong jhandle) {
  auto* metadata =
      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileMetaData*>(jhandle);
  return ROCKSDB_NAMESPACE::JniUtil::copyBytes(env, metadata->smallestkey);
}

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    largestKey
 * Signature: (J)[B
 */
jbyteArray Java_org_rocksdb_SstFileMetaData_largestKey(JNIEnv* env, jobject,
                                                       jlong jhandle) {
  auto* metadata =
      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileMetaData*>(jhandle);
  return ROCKSDB_NAMESPACE::JniUtil::copyBytes(env, metadata->largestkey);
}

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    numReadsSampled
 * Signature: (J)J
 */
jlong Java_org_rocksdb_SstFileMetaData_numReadsSampled(JNIEnv*, jobject,
                                                       jlong jhandle) {
  auto* metadata =
      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileMetaData*>(jhandle);
  return static_cast<jlong>(metadata->num_reads_sampled);
}

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    beingCompacted
 * Signature: (J)Z
 */
jboolean Java_org_rocksdb_SstFileMetaData_beingCompacted(JNIEnv*, jobject,
                                                         jlong jhandle) {
  auto* metadata =
      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileMetaData*>(jhandle);
  return static_cast<jboolean>(metadata->being_compacted);
}

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    numEntries
 * Signature: (J)J
 */
jlong Java_org_rocksdb_SstFileMetaData_numEntries(JNIEnv*, jobject,
                                                  jlong jhandle) {
  auto* metadata =
      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileMetaData*>(jhandle);
  return static_cast<jlong>(metadata->num_entries);
}

/*
 * Class:     org_rocksdb_SstFileMetaData
 * Method:    numDeletions
 * Signature: (J)J
 */
jlong Java_org_rocksdb_SstFileMetaData_numDeletions(JNIEnv*, jobject,
                                                    jlong jhandle) {
  auto* metadata =
      reinterpret_cast<ROCKSDB_NAMESPACE::SstFileMetaData*>(jhandle);
  return static_cast<jlong>(metadata->num_deletions);
}
